caravel/assets/javascripts/explorev2/components/TimeFilter.jsx

```diff
@@ -0,0 +1,125 @@
+import * as actions from '../actions/exploreActions';
+import { connect } from 'react-redux';
+
+const sinceOptions = ['1 hour ago', '12 hours ago', '1 day ago',
+  '7 days ago', '28 days ago', '90 days ago', '1 year ago'];
+const untilOptions = ['now', '1 day ago', '7 days ago',
+  '28 days ago', '90 days ago', '1 year ago'];
+
+class TimeFilter extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      sinceOptions,
+      untilOptions,
+    };
+  }
+  changeTimeColumn(timeColumnOpt) {
+    const val = (timeColumnOpt) ? timeColumnOpt.value : null;
+    this.props.actions.setTimeColumn(val);
+  }
+  changeTimeGrain(timeGrainOpt) {
+    const val = (timeGrainOpt) ? timeGrainOpt.value : null;
+    this.props.actions.setTimeGrain(val);
+  }
+  changeSince(sinceOpt) {
+    const val = (sinceOpt) ? sinceOpt.value : null;
+    this.props.actions.setSince(val);
+  }
+  changeUntil(untilOpt) {
+    const val = (untilOpt) ? untilOpt.value : null;
+    this.props.actions.setUntil(val);
+  }
+  render() {
+    return (
+      <div className="panel space-1">
+        <div className="panel-header">Time Filter</div>
+        <div className="panel-body">
+          <div className="row">
+            <h5 className="section-heading">Time Column & Grain</h5>
+            <Select
+              className="col-sm-6"
+              name="select-time-column"
+              placeholder={'Select a time column'}
+              options={this.props.timeColumnOpts}
+              value={this.props.timeColumn}
+              autosize={false}
+              onChange={this.changeTimeColumn.bind(this)}
+            />
+            <Select
+              className="col-sm-6"
+              name="select-time-grain"
+              placeholder={'Select a time grain'}
+              options={this.props.timeGrainOpts}
+              value={this.props.timeGrain}
+              autosize={false}
+              onChange={this.changeTimeGrain.bind(this)}
+            />
+          </div>
+          <div className="row">
+            <h5 className="section-heading">Since - Until</h5>
+            <Select
+              className="col-sm-6"
+              name="select-since"
+              placeholder={'Select Since Time'}
+              options={sinceOptions.map((s) => ({ value: s, label: s }))}
+              value={this.props.since}
+              autosize={false}
+              onChange={this.changeSince.bind(this)}
+            />
+            <Select
+              className="col-sm-6"
+              name="select-until"
+              placeholder={'Select Until Time'}
+              options={untilOptions.map((u) => ({ value: u, label: u }))}
+              value={this.props.until}
+              autosize={false}
+              onChange={this.changeUntil.bind(this)}
+            />
+          </div>
+        </div>
+      </div>
+    );
+  }
+}
+
+TimeFilter.propTypes = {
+  actions: React.PropTypes.object,
```

Comment:
it's a nice pattern to put proptypes at the top of the file like this:

```js
const propTypes = {
  actions: React.PropTypes.object,
  timeColumnOpts: React.PropTypes.array,
  timeColumn: React.PropTypes.string,
  timeGrainOpts: React.PropTypes.array,
  timeGrain: React.PropTypes.string,
  since: React.PropTypes.string,
  until: React.PropTypes.string,
};

const defaultProps = {
 ...
};
```

and then here at the bottom:

```js
TimeFilter.propTypes = propTypes;
TimeFilter.defaultProps = defaultProps;
```